llvm/include/llvm/TableGen/Record.h

```diff
@@ -1660,7 +1660,7 @@ class Record {
   // this record.
   SmallVector<SMLoc, 4> Locs;
   SmallVector<SMLoc, 0> ForwardDeclarationLocs;
-  SmallVector<SMRange, 0> ReferenceLocs;
+  mutable SmallVector<SMRange, 0> ReferenceLocs;
```

Comment:
You have the const_cast on the addition, so this is unnecessary?

Reply:
No, this is for another case. Essentially, `appendReferenceLoc` is only called from TGParser, and now that all those pointers are const (because in most cases, reference locs is added post creation of the record) all those functions need to either work with const pointers (by making the `ReferenceLocs` mutable) or we just const_cast<> in TGParser at the point where it's called. Either way works. Let me know what you think is better, and I will make all of them consistent.

Maybe const_cast<> so that the change is localized to the actual place?

Reply:
This was needed in `ParseIDValue`:

```
  if (Init *I = Records.getGlobal(Name->getValue())) {
    // Add a reference to the global if it's a record.
    if (TrackReferenceLocs) {
      if (auto *Def = dyn_cast<DefInit>(I)) 
        const_cast<Record *>(Def->getDef())->appendReferenceLoc(NameLoc);
    }
    return I;
  }

Def->getDef() is not a const pointer. I can change the code to what is shown above and make the function non-const.